### PR TITLE
Update action.py to have unready function

### PR DESCRIPTION
**Pull request by abant07**

### CODE REVIEW

**Comment by abant07** [comment]
Hi @Segelzwerg 

I researched unit testing in Python, but I am still unable to figure out why some lines are tested and some are not. How do I know what I should put in my tests methods for testing those specific lines? Also, where should I write these tests. Do I write it straight into the action.py file or write in test folder, if so what file in the test folder?

Thanks

**Comment by Segelzwerg** [comment]
> Hi @Segelzwerg
> 
> I researched unit testing in Python, but I am still unable to figure out why some lines are tested and some are not. How do I know what I should put in my tests methods for testing those specific lines? Also, where should I write these tests. Do I write it straight into the action.py file or write in test folder, if so what file in the test folder?
> 
> Thanks

@abant07 This is the test file for the [action module](https://github.com/Whist-Team/Whist-Server/blob/main/tests/whist/server/api/game/test_action.py). It contains the test for the `ready` which should clear all your questions. Please don't open multiple PRs for the same feature.

**Comment by abant07** [comment]
Hi @Segelzwerg 

So, I added to the test file, and everything seems to working except I am getting python application errors. So, I tried looking at the details, but I am not able to understand what it saying and what the errors are. I have pasted a snippet of an example of the errors.

self = <tests.whist.server.api.game.test_action.ActionGameTestCase testMethod=test_unready_game_not_found>

    def test_unready_game_not_found(self):
        self.game_mock.unready_player = MagicMock(side_effect=GameNotFoundError)
>       response = self.client.post(url=f'/game/action/unready/{self.game_mock.id}',
                                    headers=self.second_player)

tests/whist/server/api/game/test_action.py:85: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/opt/hostedtoolcache/Python/3.9.12/x64/lib/python3.9/site-packages/requests/sessions.py:577: in post
    return self.request('POST', url, data=data, json=json, **kwargs)
/opt/hostedtoolcache/Python/3.9.12/x64/lib/python3.9/site-packages/starlette/testclient.py:476: in request
    return super().request(
/opt/hostedtoolcache/Python/3.9.12/x64/lib/python3.9/site-packages/requests/sessions.py:529: in request
    resp = self.send(prep, **send_kwargs)
/opt/hostedtoolcache/Python/3.9.12/x64/lib/python3.9/site-packages/requests/sessions.py:645: in send
    r = adapter.send(request, **kwargs)
/opt/hostedtoolcache/Python/3.9.12/x64/lib/python3.9/site-packages/starlette/testclient.py:270: in send
    raise exc
/opt/hostedtoolcache/Python/3.9.12/x64/lib/python3.9/site-packages/starlette/testclient.py:267: in send
    portal.call(self.app, scope, receive, send)
/opt/hostedtoolcache/Python/3.9.12/x64/lib/python3.9/site-packages/anyio/from_thread.py:283: in call
    return cast(T_Retval, self.start_task_soon(func, *args).result())
/opt/hostedtoolcache/Python/3.9.12/x64/lib/python3.9/concurrent/futures/_base.py:446: in result
    return self.__get_result()
/opt/hostedtoolcache/Python/3.9.12/x64/lib/python3.9/concurrent/futures/_base.py:391: in __get_result
    raise self._exception
/opt/hostedtoolcache/Python/3.9.12/x64/lib/python3.9/site-packages/anyio/from_thread.py:219: in _call_func
    retval = await retval
/opt/hostedtoolcache/Python/3.9.12/x64/lib/python3.9/site-packages/fastapi/applications.py:269: in __call__
    await super().__call__(scope, receive, send)
/opt/hostedtoolcache/Python/3.9.12/x64/lib/python3.9/site-packages/starlette/applications.py:124: in __call__
    await self.middleware_stack(scope, receive, send)
/opt/hostedtoolcache/Python/3.9.12/x64/lib/python3.9/site-packages/starlette/middleware/errors.py:184: in __call__
    raise exc
/opt/hostedtoolcache/Python/3.9.12/x64/lib/python3.9/site-packages/starlette/middleware/errors.py:162: in __call__
    await self.app(scope, receive, _send)
/opt/hostedtoolcache/Python/3.9.12/x64/lib/python3.9/site-packages/starlette/exceptions.py:93: in __call__
    raise exc
/opt/hostedtoolcache/Python/3.9.12/x64/lib/python3.9/site-packages/starlette/exceptions.py:82: in __call__
    await self.app(scope, receive, sender)
/opt/hostedtoolcache/Python/3.9.12/x64/lib/python3.9/site-packages/fastapi/middleware/asyncexitstack.py:21: in __call__
    raise e
/opt/hostedtoolcache/Python/3.9.12/x64/lib/python3.9/site-packages/fastapi/middleware/asyncexitstack.py:18: in __call__
    await self.app(scope, receive, send)
/opt/hostedtoolcache/Python/3.9.12/x64/lib/python3.9/site-packages/starlette/routing.py:670: in __call__
    await route.handle(scope, receive, send)
/opt/hostedtoolcache/Python/3.9.12/x64/lib/python3.9/site-packages/starlette/routing.py:266: in handle
    await self.app(scope, receive, send)
/opt/hostedtoolcache/Python/3.9.12/x64/lib/python3.9/site-packages/starlette/routing.py:65: in app
    response = await func(request)
/opt/hostedtoolcache/Python/3.9.12/x64/lib/python3.9/site-packages/fastapi/routing.py:227: in app
    raw_response = await run_endpoint_function(
/opt/hostedtoolcache/Python/3.9.12/x64/lib/python3.9/site-packages/fastapi/routing.py:162: in run_endpoint_function
    return await run_in_threadpool(dependant.call, **values)
/opt/hostedtoolcache/Python/3.9.12/x64/lib/python3.9/site-packages/starlette/concurrency.py:41: in run_in_threadpool
    return await anyio.to_thread.run_sync(func, *args)
/opt/hostedtoolcache/Python/3.9.12/x64/lib/python3.9/site-packages/anyio/to_thread.py:31: in run_sync
    return await get_asynclib().run_sync_in_worker_thread(
/opt/hostedtoolcache/Python/3.9.12/x64/lib/python3.9/site-packages/anyio/_backends/_asyncio.py:9[37](https://github.com/Whist-Team/Whist-Server/runs/6628627967?check_suite_focus=true#step:8:38): in run_sync_in_worker_thread
    return await future
/opt/hostedtoolcache/Python/3.9.12/x[64](https://github.com/Whist-Team/Whist-Server/runs/6628627967?check_suite_focus=true#step:8:65)/lib/python3.9/site-packages/anyio/_backends/_asyncio.py:8[67](https://github.com/Whist-Team/Whist-Server/runs/6628627967?check_suite_focus=true#step:8:68): in run
    result = context.run(func, *args)
whist/server/api/game/action.py:126: in unready_player
    game.unready_player(user)
/opt/hostedtoolcache/Python/3.9.12/x64/lib/python3.9/unittest/mock.py:10[92](https://github.com/Whist-Team/Whist-Server/runs/6628627967?check_suite_focus=true#step:8:93): in __call__
    return self._mock_call(*args, **kwargs)
/opt/hostedtoolcache/Python/3.9.12/x64/lib/python3.9/unittest/mock.py:10[96](https://github.com/Whist-Team/Whist-Server/runs/6628627967?check_suite_focus=true#step:8:97): in _mock_call
    return self._execute_mock_call(*args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <MagicMock name='mock.unready_player' id='139884536775920'>
args = (<MagicMock name='marcel' id='139884536760736'>,), kwargs = {}
effect = <class 'tests.whist.server.api.game.test_action.GameNotFoundError'>

    def _execute_mock_call(self, /, *args, **kwargs):
        # separate from _increment_mock_call so that awaited functions are
        # executed separately from their call, also AsyncMock overrides this method
    
        effect = self.side_effect
        if effect is not None:
            if _is_exception(effect):
>               raise effect
E               tests.whist.server.api.game.test_action.GameNotFoundError

/opt/hostedtoolcache/Python/3.9.12/x64/lib/python3.9/unittest/mock.py:[115](https://github.com/Whist-Team/Whist-Server/runs/6628627967?check_suite_focus=true#step:8:116)1: GameNotFoundError

File: tests/whist/server/api/game/test_action.py
Comment (Segelzwerg) [inline tests/whist/server/api/game/test_action.py:21]:
```suggestion
```

@abant07  I suspect the error is that you redeclare the errors here, but you need to import them here.

File: tests/whist/server/api/game/test_action.py
Comment (Segelzwerg) [inline tests/whist/server/api/game/test_action.py:8]:
```suggestion
from whist.server.services.error import GameNotFoundError
from whist.server.services.game_db_service import GameDatabaseService

```

File: whist/server/api/game/action.py
Comment (Segelzwerg) [inline whist/server/api/game/action.py:28]: The errors shouldn't be defined here. `GameNotFoundError` is already defined somewhere else. See above.

**Review by Segelzwerg — CHANGES_REQUESTED** [review]

**Comment by iTitus** [comment]
This branch is also out-of-date

**Comment by abant07** [comment]
Hi @Segelzwerg 

Thank you for all the help on this issue! Are there any other revisions that you see that need to be made to the code?

Thanks

File: whist/server/api/game/action.py
Comment (Segelzwerg) [inline whist/server/api/game/action.py:126]:
```suggestion
        message = 'Game not found'
```

File: whist/server/api/game/action.py
Comment (Segelzwerg) [inline whist/server/api/game/action.py:133]:
```suggestion
        message = 'Player not ready'
```

File: whist/server/api/game/create.py
Comment (Segelzwerg) [inline whist/server/api/game/create.py:5]:
```suggestion
```
This is out of scope of this PR.

File: whist/server/api/game/create.py
Comment (Segelzwerg) [inline whist/server/api/game/create.py:9]:
```suggestion
```

File: whist/server/api/game/create.py
Comment (Segelzwerg) [inline whist/server/api/game/create.py:13]:
```suggestion
```

File: whist/server/api/game/create.py
Comment (Segelzwerg) [inline whist/server/api/game/create.py:32]:
```suggestion
```

File: whist/server/api/game/join.py
Comment (Segelzwerg) [inline whist/server/api/game/join.py:7]:
```suggestion
```

File: whist/server/api/game/join.py
Comment (Segelzwerg) [inline whist/server/api/game/join.py:17]:
```suggestion
```

File: whist/server/api/game/join.py
Comment (Segelzwerg) [inline whist/server/api/game/join.py:48]:
```suggestion
```

File: whist/server/api/game/join.py
Comment (Segelzwerg) [inline whist/server/api/game/join.py:50]:
```suggestion
```

File: whist/server/api/user/auth.py
Comment (Segelzwerg) [inline whist/server/api/user/auth.py:4]:
```suggestion
```

File: whist/server/api/user/auth.py
Comment (Segelzwerg) [inline whist/server/api/user/auth.py:13]:
```suggestion
```

File: whist/server/api/user/auth.py
Comment (Segelzwerg) [inline whist/server/api/user/auth.py:42]:
```suggestion
```

File: whist/server/api/user/create.py
Comment (Segelzwerg) [inline whist/server/api/user/create.py:5]:
```suggestion
```

File: whist/server/api/user/create.py
Comment (Segelzwerg) [inline whist/server/api/user/create.py:13]:
```suggestion
```

File: whist/server/api/user/create.py
Comment (Segelzwerg) [inline whist/server/api/user/create.py:40]:
```suggestion
```

File: whist/server/database/game.py
Comment (Segelzwerg) [inline whist/server/database/game.py:3]:
```suggestion
```

File: whist/server/services/error.py
Comment (Segelzwerg) [inline whist/server/services/error.py:16]:
```suggestion
    Is raised when an user is not to start.
```

**Review by Segelzwerg — CHANGES_REQUESTED** [review]
Other than the clean-up LGTM. Thank you very much.

After you apply the changes I'll merge.

File: whist/server/api/game/create.py
Comment (abant07) [reply whist/server/api/game/create.py:7]: if I take out this import, errors start to occur, so should I still take it out?

**Comment by abant07** [comment]
Hi @Segelzwerg 

I finished with all the review changes from this issue, so I believe it can be ready to merge. Thank you!

I also wanted to see if the code I did for API Logging was merged, since I finished it a while back, and don't see the code updated in the server's "whist-api" folders.

File: whist/server/api/game/create.py
Comment (Segelzwerg) [reply whist/server/api/game/create.py:7]: yes you are right, I missed that line

File: whist/server/database/game.py
Comment (Segelzwerg) [inline whist/server/database/game.py:2]:
```suggestion
```

**Review by Segelzwerg — CHANGES_REQUESTED** [review]
Sorry I missed one empty line. Otherwise is complete

File: whist/server/api/game/create.py
Comment (Segelzwerg) [inline whist/server/api/game/create.py:9]:
```suggestion
from whist.server.services.authentication import get_current_user
from whist.server.services.game_db_service import GameDatabaseService
from whist.server.services.password import PasswordService
```

sorry my bad again. I suggested the wrong order. We try to keep otherwise unmodified files untouched.

**Review by Segelzwerg — CHANGES_REQUESTED** [review]

**Review by Segelzwerg — APPROVED** [review]

**Pull request merged**
